### PR TITLE
Add author to post item

### PR DESCRIPTION
**Pull request by mjuhasz**
This PR adds the author to the post item card if the site is a multi-user site, or All Sites mode is selected with posts from multiple users.

All Sites mode, multiple authors:
screen shot 2017-08-25 at 10 56 26 pm

Single site mode, multiple authors:

screen shot 2017-08-25 at 11 03 09 pm

Single site mode, single author:

screen shot 2017-08-25 at 11 04 23 pm

This PR is part of a larger epic and behind the posts/post-type-list feature flag.
See p8F9tW-hL-p2.

To test:

- Check out this branch
- ENABLE_FEATURES=posts/post-type-list npm start
- Navigate to Blog Posts
- Select All Sites
- Verify that the author is visible if posts are from multiple authors
- Verify that the Site icon and site name are visible
- Select a specific site
- Verify that the author is visible if the site has multiple authors
- Verify that the Site icon and site name are not visible

### CODE REVIEW

File: client/blocks/post-item/index.jsx [hunk 1]
```diff
@@ -171,14 +182,10 @@ class PostItem extends React.Component {
...
-									<SiteIcon size={ 16 } site={ site } />
-									<div className="post-item__site-title">
-										{ siteTitle }
-									</div>
-								</div>
-							) }
+							<div className="post-item__info">
+								<PostTypeSiteInfo globalId={ globalId } />
+								<PostTypePostAuthor globalId={ globalId } />
+							</div>
```
Comment (nylen) [inline client/blocks/post-item/index.jsx]: We already have variables here called `isAuthorVisible` and `isSiteVisible` (better: `isSiteInfoVisible`).  What about using these variables to control whether these components are rendered here?  This should allow us to remove the dependency on these same parameters in the sub-components, for example [here](https://github.com/Automattic/wp-calypso/pull/17549/files#diff-a382ca2105ae121973f7a063fdea4857R19).

File: client/my-sites/post-type-list/post-type-post-author/index.jsx [hunk 2]
```diff
@@ -1,38 +1,43 @@
...
+import { areAllSitesSingleUser } from 'state/selectors';
+import { getSelectedSiteId } from 'state/ui/selectors';
 
-function PostTypePostAuthor( { translate, singleUserSite, name } ) {
-	if ( ! name || singleUserSite ) {
+function PostTypePostAuthor( { singleUserSite, isAllSitesModeSelected, allSitesSingleUser, name } ) {
+	if ( ! isEnabled( 'posts/post-type-list' ) ||
+		! name ||
+		( isAllSitesModeSelected && allSitesSingleUser ) ||
+		( ! isAllSitesModeSelected && singleUserSite ) ) {
```
Comment (nylen) [inline client/my-sites/post-type-list/post-type-post-author/index.jsx]: As noted above, I think we can avoid depending on `isEnabled`, `singleUserSite`, `isAllSitesModeSelected`, and `allSitesSingleUser` here, by allowing the parent component to decide whether to render this component.

`if ( ! name )` - is this condition possible / likely / easy to reproduce?  When that happens, should this component return a placeholder instead?

File: client/my-sites/post-type-list/post-type-site-info/index.jsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import { isEnabled } from 'config';
+import { getPost } from 'state/posts/selectors';
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getSite, getSiteTitle } from 'state/sites/selectors';
+import SiteIcon from 'blocks/site-icon';
+
+function PostTypeSiteInfo( { site, siteTitle, isAllSitesModeSelected } ) {
+	if ( ! isEnabled( 'posts/post-type-list' ) || ! site || ! isAllSitesModeSelected ) {
+		return null;
+	}
```
Comment (nylen) [inline client/my-sites/post-type-list/post-type-site-info/index.jsx]: As noted above, I think we can avoid depending on `isEnabled` and `isAllSitesModeSelected` here, by allowing the parent component to decide whether to render this component.

`if ( ! site )` - is this condition possible / likely / easy to reproduce?  When that happens, should this component return a placeholder instead?

File: client/blocks/post-item/index.jsx (hunk 1 above)
Comment (mjuhasz) [reply client/blocks/post-item/index.jsx]: Good point, I just pushed a commit to change that!

File: client/my-sites/post-type-list/post-type-post-author/index.jsx (hunk 2 above)
Comment (mjuhasz) [reply client/my-sites/post-type-list/post-type-post-author/index.jsx]: I just pushed a commit to allow the parent to decide about the visibility of the child component.

File: client/my-sites/post-type-list/post-type-site-info/index.jsx (hunk 3 above)
Comment (mjuhasz) [reply client/my-sites/post-type-list/post-type-site-info/index.jsx]: I just pushed a commit to allow the parent to decide about the visibility of the child component.

`if ( ! site )` is `true` while the Blog Posts page is being loaded in All Sites mode. For me, that condition is hit 15-20 times before the site becomes known. So it's possible, and easy to reproduce. The post item card already looks like a placeholder while the post list is being loaded so visually I don't see any issue without rendering a placeholder here explicitly.

File: client/my-sites/post-type-list/post-type-post-author/index.jsx (hunk 2 above)
Comment (mjuhasz) [reply client/my-sites/post-type-list/post-type-post-author/index.jsx]: I haven't been able to reproduce `if ( ! name )` but that condition was there before so there may be circumstances where it is necessary. I don't think we need a placeholder (see my comment about site name.)

**Comment by nylen** [comment]
I think the author field should be hidden on the `/posts/my` and `/posts/my/:site` views - by definition, they are only showing posts by a single author:

>

**Comment by nylen** [comment]
This is more general feedback, and I believe in this case, this PR is working as intended.  However, it's not always clear what an author display is on these cards.  Here's an example with author info shown and site info hidden (again, as expected):

>

**Comment by nylen** [comment]
There are some wrapping / height issues on small screen widths when all info for a post is shown:

> 

This should be fixed either by #17550 or by removing `react-virtualized` from `PostTypeList`.

**Comment by nylen** [comment]
Except for the issues noted above, this is working correctly.

**Comment by nylen** [comment]
> I think the author field should be hidden on the `/posts/my` and `/posts/my/:site` views - by definition, they are only showing posts by a single author

This is fixed in 44b487a.  I think the other issues can be addressed separately, so I am marking this PR as ready to merge; feel free to revise if you'd like someone else to take a look first.

**Pull request merged**